boa/src/builtins/json/mod.rs

```diff
@@ -62,10 +66,73 @@ pub fn parse(_: &mut Value, args: &[Value], _: &mut Interpreter) -> ResultValue
 ///
 /// [spec]: https://tc39.es/ecma262/#sec-json.stringify
 /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify
-pub fn stringify(_: &mut Value, args: &[Value], _: &mut Interpreter) -> ResultValue {
-    let obj = args.get(0).expect("cannot get argument for JSON.stringify");
-    let json = obj.to_json().to_string();
-    Ok(Value::from(json))
+pub fn stringify(_: &mut Value, args: &[Value], interpreter: &mut Interpreter) -> ResultValue {
+    let object = args.get(0).expect("cannot get argument for JSON.stringify");
+
+    if args.len() == 1 {
+        return Ok(Value::from(object.to_json().to_string()));
+    }
+
+    let replacer = args.get(1).expect("cannot get JSON.stringify replacer");
+    if !replacer.is_object() {
+        return Ok(Value::from(object.to_json().to_string()));
+    }
+
+    let replacer_as_object = replacer
+        .as_object()
+        .expect("JSON.stringify replacer was an object");
+    if replacer_as_object.is_callable() {
+        object
+            .as_object()
+            .map(|obj| {
+                let object_to_return = Value::new_object(None);
+                for (key, val) in obj.properties.iter() {
+                    if let Some(value) = &val.value {
+                        let mut this_arg = object.clone();
+                        object_to_return.set_property(
+                            String::from(key),
+                            Property::default().value(
+                                interpreter
+                                    .call(
+                                        replacer,
+                                        &mut this_arg,
+                                        &[Value::string(key), Value::from(value)],
+                                    )
+                                    .expect("failed to get returned value from replacer function"),
+                            ),
+                        );
+                    }
+                }
+                Value::from(object_to_return.to_json().to_string())
+            })
+            .ok_or_else(Value::undefined)
+    } else if replacer_as_object.kind == ObjectKind::Array {
+        let mut obj_to_return =
+            serde_json::Map::with_capacity(replacer_as_object.properties.len() - 1);
+        let fields = replacer_as_object
+            .properties
+            .iter()
+            .filter_map(|(key, prop)| {
+                if key == "length" {
+                    None
+                } else {
+                    // TODO: this should be the abstract operation `Get`
+                    prop.value.as_ref().map(Value::to_string)
```

Comment:
We should change this to do `get_field()` as we do in `to_json()`, and iterate only over the keys.